File: appwin_analytics/ios/appwin_analytics/Sources/appwin_analytics/AppwinAnalyticsPlugin.swift
import Flutter
import UIKit
import AppwinCore
import AppwinAnalytics

// Native modules are declared as `s.dependency` in the podspec (ADR-0020,
// Firebase/FlutterFire pattern). This file is the only glue; the
// implementations live in AppwinCore and AppwinAnalytics.
public class AppwinAnalyticsPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "appwin_analytics", binaryMessenger: registrar.messenger())
        let instance = AppwinAnalyticsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    @MainActor
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "initialize":
            // Availability, not configuration: `AppwinCore.configure` is the
            // host app's job and runs through the appwin_core plugin.
            Task {
                let verdict = await AppwinAnalytics.initialize()
                await MainActor.run { result(Self.encode(verdict)) }
            }
        case "track":
            guard let args = call.arguments as? [String: Any],
                let name = args["name"] as? String
            else {
                result(FlutterError(code: "bad_args", message: "name manquant", details: nil))
                return
            }
            AppwinAnalytics.track(name, props: Self.parseProps(args["props"]))
            result(nil)
        case "screen":
            guard let args = call.arguments as? [String: Any],
                let name = args["name"] as? String
            else {
                result(FlutterError(code: "bad_args", message: "name manquant", details: nil))
                return
            }
            AppwinAnalytics.screen(name)
            result(nil)
        case "flush":
            AppwinAnalytics.flush()
            result(nil)
        case "setConsent":
            let raw = (call.arguments as? [String: Any])?["consent"] as? String
            AppwinAnalytics.setConsent(Self.parseConsent(raw))
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /// Method-channel props to the SDK's typed values. Numbers cross the
    /// channel as NSNumber: booleans are told apart via the ObjC type
    /// encoding, ints stay ints, everything decimal becomes a double.
    private static func parseProps(_ raw: Any?) -> [String: AnalyticsValue]? {
        guard let dict = raw as? [String: Any], !dict.isEmpty else { return nil }
        var out: [String: AnalyticsValue] = [:]
        for (key, value) in dict {
            switch value {
            case let number as NSNumber:
                if CFGetTypeID(number) == CFBooleanGetTypeID() {
                    out[key] = .bool(number.boolValue)
                } else if CFNumberIsFloatType(number) {
                    out[key] = .double(number.doubleValue)
                } else {
                    out[key] = .int(number.intValue)
                }
            case let string as String:
                out[key] = .string(string)
            default:
                out[key] = AnalyticsValue.string(String(describing: value))
            }
        }
        return out
    }

    private static func parseConsent(_ raw: String?) -> AnalyticsConsent {
        switch raw {
        case "granted": return .granted
        case "denied": return .denied
        default: return .unknown
        }
    }

    /// Maps the native result onto what the Dart side parses. Same shape
    /// as every other product plugin: status + optional reason.
    private static func encode(_ verdict: AppwinInitResult) -> [String: Any] {
        switch verdict {
        case .ready:
            return ["status": "ready"]
        case .notConfigured:
            return ["status": "notConfigured"]
        case .unknown:
            return ["status": "unknown"]
        case .unavailable(let reason):
            return ["status": "unavailable", "reason": reason.rawValue]
        }
    }
}

File: appwin_analytics/android/src/main/kotlin/io/appwin/flutter/analytics/AppwinAnalyticsPlugin.kt
package io.appwin.flutter.analytics

import android.os.Build
import io.appwin.analytics.AnalyticsConsent
import io.appwin.analytics.AppwinAnalytics
import io.appwin.core.availability.AppwinInitResult
import io.appwin.core.availability.AppwinInitStatus
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Dart-to-Kotlin glue for the Analytics product.
 *
 * Mirrors `AppwinAnalyticsPlugin.swift`: same method names, same arguments,
 * same result shapes. One Dart implementation serves both platforms, so any
 * drift here is paid by the studio that only tests on one phone.
 */
class AppwinAnalyticsPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

  private var channel: MethodChannel? = null
  private var scope: CoroutineScope? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    channel = MethodChannel(binding.binaryMessenger, "appwin_analytics").also {
      it.setMethodCallHandler(this)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setMethodCallHandler(null)
    channel = null
    scope?.cancel()
    scope = null
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${Build.VERSION.RELEASE}")
      "initialize" -> {
        val scope = scope ?: return result.error("not_attached", "plugin detached", null)
        scope.launch {
          result.success(encode(AppwinAnalytics.initialize()))
        }
      }
      "track" -> {
        val name = call.argument<String>("name")
          ?: return result.error("bad_args", "name manquant", null)
        AppwinAnalytics.track(name, call.argument<Map<String, Any?>>("props"))
        result.success(null)
      }
      "screen" -> {
        val name = call.argument<String>("name")
          ?: return result.error("bad_args", "name manquant", null)
        AppwinAnalytics.screen(name)
        result.success(null)
      }
      "flush" -> {
        AppwinAnalytics.flush()
        result.success(null)
      }
      "setConsent" -> {
        AppwinAnalytics.setConsent(parseConsent(call.argument<String>("consent")))
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  private fun parseConsent(raw: String?): AnalyticsConsent = when (raw) {
    "granted" -> AnalyticsConsent.GRANTED
    "denied" -> AnalyticsConsent.DENIED
    else -> AnalyticsConsent.UNKNOWN
  }

  /** Same shape as the iOS plugin: status + optional reason. */
  private fun encode(result: AppwinInitResult): Map<String, Any?> = when (result.status) {
    AppwinInitStatus.READY -> mapOf("status" to "ready")
    AppwinInitStatus.NOT_CONFIGURED -> mapOf("status" to "notConfigured")
    AppwinInitStatus.UNKNOWN -> mapOf("status" to "unknown")
    AppwinInitStatus.UNAVAILABLE ->
      mapOf("status" to "unavailable", "reason" to result.reason?.key)
  }
}
